src/main/java/com/microsoft/graph/tasks/PageIterator.java

```diff
@@ -0,0 +1,494 @@
+import com.microsoft.graph.exceptions.ServiceException;
+import com.microsoft.graph.requests.IBaseClient;
+import com.microsoft.kiota.HttpMethod;
+import com.microsoft.kiota.RequestAdapter;
+import com.microsoft.kiota.RequestInformation;
+import com.microsoft.kiota.serialization.AdditionalDataHolder;
+import com.microsoft.kiota.serialization.Parsable;
+import com.microsoft.kiota.serialization.ParsableFactory;
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.util.*;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Function;
+import java.util.function.UnaryOperator;
+
+/**
+ * A class for iterating through pages of a collection
+ * Uses to automatically pages through result sets across multiple calls and process each item in the result set.
+ * @param <TEntity> The type of the entity returned in the collection. This type must implement {@link Parsable}
+ * @param <TCollectionPage> The Microsoft Graph collection response type returned in the collection response. This type must implement {@link Parsable} and {@link AdditionalDataHolder}
+ */
+public class PageIterator<TEntity extends Parsable, TCollectionPage extends Parsable & AdditionalDataHolder> {
+    /**
+     * Creates a new instance of the PageIterator class
+     */
+    protected PageIterator() {
+        // default constructor
+    }
+    private static final String NO_COLLECTION_PROPERTY_ERROR = "The Parsable does not contain a collection property.";
+    private RequestAdapter requestAdapter;
+    private TCollectionPage currentPage;
+    private ParsableFactory<TCollectionPage> collectionPageFactory;
+    private Queue<TEntity> pageItemQueue;
+    private Function<TEntity, Boolean> processPageItemCallback;
+    private Function<TEntity, CompletableFuture<Boolean>> asyncProcessPageItemCallback;
+    private UnaryOperator<RequestInformation> requestConfigurator;
+
+
+    private String deltaLink;
+
+    /**
+     * The deltaLink returned from a delta query.
+     * @return the deltaLink from the delta query
+     */
+    @Nullable
+    public String getDeltaLink() {
+        return deltaLink;
+    }
+    private String nextLink;
+    /**
+     * The nextLink returned from a collection query.
+     * @return the nextLink from the collection query
+     */
+    @Nullable
+    public String getNextLink() {
+        return nextLink;
+    }
+    private PageIteratorState state;
+    /**
+     * The state of the page iterator
+     * @return the state of the page iterator
+     */
+    @Nonnull
+    public PageIteratorState getPageIteratorState() {
+        return state;
+    }
+    /**
+     * Set the state of the page iterator
+     * @param state the state of the page iterator
+     */
+    public void setPageIteratorState(@Nonnull PageIteratorState state) {
```

Comment:
I don't think we should have this one to be available publicly as it calculated. Maybe we can remove this setter and have the property initialized in the default constructor?